packages/router/router.ts

```diff
@@ -191,6 +192,64 @@ export interface Router {
    */
   dispose(): void;
 
+  /**
+   * Create a new navigation blocker
+   *
+   * @param key A unique identifier for the blocker
+   */
+  createBlocker(key: string, fn: ShouldBlockFunction): Blocker;
+
+  /**
+   * Get a navigation blocker
+   *
+   * @param key The identifier for the blocker
+   */
+  getBlocker(key: string): Blocker | undefined;
+
+  /**
+   * Delete a navigation blocker
+   *
+   * @param key The identifier for the blocker
+   */
+  deleteBlocker(key: string): void;
+
+  /**
+   * Update the state of a navigation blocker
+   *
+   * @param key The identifier for the blocker
+   * @param state The next state. Must be one of `"blocked"`, `"unblocked"`, or
+   * "proceeding"
+   * @param navigate The navigation function to call when the blocked navigation
+   * is allowed to proceed
+   */
+  setBlockerState(
+    key: string,
+    state: "blocked",
+    navigate: () => Promise<void>
+  ): Blocker | undefined;
+
+  /**
+   * Update the state of a navigation blocker
+   *
+   * @param key The identifier for the blocker
+   * @param state The next state. Must be one of `"blocked"`, `"unblocked"`, or
+   * "proceeding"
+   */
+  setBlockerState(
+    key: string,
+    state: "unblocked" | "proceeding"
+  ): Blocker | undefined;
+
+  /**
+   * Update a navigation blocker's function. This is the function that is
+   * evaluated to determine whether or not to block a navigation event.
+   *
+   * @param key The identifier for the blocker
+   * @param fn The updated function to determine whether or not to block a
+   * navigation
+   */
+  setBlockerFunction(key: string, fn: ShouldBlockFunction): Blocker | undefined;
+
```

Comment:
I wonder if we could trim this router public API - the more we expose to be called from the react layer the more code we have to write in the preact/vue layers etc.

I wonder if this could mimic the fetcher api?  It might need one more entry point since we're providing behavior upon creation, but maybe we could get away with function identity checks?


```js
function useBlocker(fn) {
  let key = // use ref or whatever to get a stable key
  let callback = useCallback(() => fn(), [fn])
  let blocker = router.getBlocker(key, callback);
  //                    ^ internally this would create a new blocker if it didn't yet exist, 
  //                      and would update the function if it was different

  useEffect(() => {
    return () => router.deleteBlocker(key);
  }, [router, key]);

  return blocker;
}
```

Thoughts?